### ISSUE TITLE
B018 does not detect unused tuple expressions

### ISSUE THREAD

**Issue by LordAro** [astral-sh/ruff#14131]
B018 (useless-expression) does not detect function calls with bare tuples:

```python
def foo(): ...

foo(),
```

COM818 (trailing-comma-on-bare-tuple) detects this as it should, but I feel like B018 should flag this as well?

Ruff 0.7.2

**Comment by MichaReiser** [comment]
Flagging this as `B018` is only correct if `foo` has no side effects. For example, you could have

```py
print("test"),
```

Removing the `print` would change the program behavior, which is why it's not useless.

Reasoning whether a function is only be possible in the most trivial cases because standard library functions in typeshed don't annotate if they have side effects.

**Comment by LordAro** [comment]
Indeed, and a `print(),` is originally how I found this behaviour.

Maybe the answer is just to enable flake8-commas linting...

**Comment by LordAro** [comment]
Actually, hmm. It's definitely still a useless expression, just a weirdly formatted one where the "fix" is to remove the `,` rather than the entire expression

**Comment by MichaReiser** [comment]
Yeah. I think detecting unnecessary tuples could be in the spirit of the rule, but it could also be its own rule, similar to `useless-comparison.`

**Comment by MichaReiser** [comment]
I think the idea here is that `print("hello"),` would be fixed to just `print("hello")`, because the tuple expression is useless, not its content. But that might be difficult to convey in a diagnostic.

> On the other hand, as you point out, we have a lint rule detecting trailing commas on bare tuples. So I'd think it would be the job of that lint rule to fix this expression.

I should have taken a closer look at that rule. I agree, we already have a rule that covers what you want.

**Comment by LordAro** [comment]
Ah, ok, I've properly understood what I want: [pylint W0106](https://pylint.readthedocs.io/en/stable/user_guide/messages/warning/expression-not-assigned.html) which *does* warn about such things

```python
def foo(): ...
[1, 2, foo()]
print("foo"),
```
```
test.py:1:0: W0106: Expression "[1, 2, foo()]" is assigned to nothing (expression-not-assigned)
test.py:2:0: W0106: Expression "(print('foo'), )" is assigned to nothing (expression-not-assigned)
```

obviously the fix is not to remove it entirely (as functions may have side effects) - it's one of
```
_ = [1, 2, foo()]  # a
foo()  # b
```

I don't think you'll need a type checker to do that (though pyright does detect the unused expression)

```
  /foo/test.py:1:1 - error: Expression value is unused (reportUnusedExpression)
  /foo/test.py:2:1 - error: Expression value is unused (reportUnusedExpression)
```

in #970 this lint is listed as implemented via B018, so...

**Comment by dylwil3** [comment]
Ok thanks this helps me understand better!

My takeaway is that `W0106` is a slightly different rule that should just be implemented on its own, and would maybe suggest the fix "a". 

I still feel like "useless" indicates "can be removed" not "should be modified", so I would vote to keep B018 from emitting a diagnostic for this case, but then you have `COM818` and a newly implemented `W0106` to fall back on. Does that seem ok? I'm still open to being convinced out of my pedantic interpretation of the word "useless"!

**Comment by LordAro** [comment]
I mean, if you look at actual flake8-bugbear:

>B018: Found useless expression. Either assign it to a variable or remove it. Note that dangling commas will cause things to be interpreted as useless tuples. For example, in the statement `print(".."),` is the same as `(print(".."),)` which is an unassigned tuple. Simply remove the comma to clear the error.

**Comment by dylwil3** [comment]
Alright, I stand corrected! We should certainly have parity with bugbear for this. Thanks for your persistence in the face of my pedantry!

**Comment by dylwil3** [comment]
Actually, despite the apparent resolution above, I have had some conversations and am again thinking that this should not be implemented. I'll have a few more conversations and get back to you if we end up still going forward with this- thanks!

**Comment by ntBre** [comment]
After looking at this while reviewing https://github.com/astral-sh/ruff/pull/25318, I'm putting this back to rule+needs-decision. There was this interesting case in the ecosystem check:
+ [providers/openlineage/tests/unit/openlineage/plugins/test_utils.py:82:17:](https://github.com/apache/airflow/blob/91843291f540aed5c3de70148d53b299b08e48f6/providers/openlineage/tests/unit/openlineage/plugins/test_utils.py#L82) 

where it looked like the tuple was being used intentionally to keep two expressions on the same line:

```py
class SafeStrDict(dict):
    def __str__(self):
        castable = []
        for key, val in self.items():
            try:
                str(key), str(val)
                castable.append((key, val))
            except (TypeError, NotImplementedError):
                continue
        return str(dict(castable))
```

That kind of pushes me toward making this a separate rule that covers the tuple case and possibly the list case as well (`useless-collection`?). Maybe I'm reading too much into this, and they should use a semi-colon or just two separate lines instead, though. The other cases looked like true positives.

I'm also fine with including it in `B018` (in preview), but I feel like we should handle lists and tuples together one way or another like pylint and bugbear.
